File: 02. Estructurales/05. Facade/01. facade.js
/***
 
Imaginemos que estamos construyendo una aplicación de compra en línea. 
Tenemos diferentes subsistemas que manejan diferentes aspectos de la compra, 
como el inventario, el carrito de compras y el procesamiento de pagos. 
Queremos simplificar la interfaz para los clientes y proporcionar una 
única interfaz fácil de usar para realizar una compra.

Aquí es donde el patrón Facade puede ser útil. 
Podemos utilizar el patrón Facade para crear una fachada que envuelva los subsistemas 
y proporcione una interfaz simplificada para realizar una compra.

*/

// Comencemos definiendo nuestros subsistemas:

// Subsistema de inventario
class InventorySystem {
    checkAvailability(item) {
        console.log(`Verificando disponibilidad de ${item} en inventario...`);
        // Lógica para verificar la disponibilidad en el inventario
        return true;
    }
}

// Subsistema de carrito de compras
class ShoppingCart {
    addItem(item) {
        console.log(`Agregando ${item} al carrito de compras...`);
        // Lógica para agregar un elemento al carrito de compras
    }

    removeItem(item) {
        console.log(`Eliminando ${item} del carrito de compras...`);
        // Lógica para eliminar un elemento del carrito de compras
    }
}

// Subsistema de procesamiento de pagos
class PaymentProcessor {
    processPayment(amount) {
        console.log(`Procesando pago de ${amount}...`);
        // Lógica para procesar el pago
        return true;
    }
}

/***
 
A continuación, definimos nuestra fachada que envuelve los subsistemas 
y proporciona una interfaz simplificada para realizar una compra:

*/

class PurchaseFacade {
    constructor() {
        this.inventorySystem = new InventorySystem();
        this.shoppingCart = new ShoppingCart();
        this.paymentProcessor = new PaymentProcessor();
    }

    purchaseItem(item, amount) {
        if (this.inventorySystem.checkAvailability(item)) {
            this.shoppingCart.addItem(item);
            if (this.paymentProcessor.processPayment(amount)) {
                console.log(`Compra exitosa: ${item}`);
            } else {
                console.log(`Error al procesar el pago`);
            }
        } else {
            console.log(`El artículo no está disponible en inventario`);
        }
    }
}

/***

    Finalmente, podemos utilizar el patrón Facade para realizar una compra de manera simplificada:

*/

const purchaseFacade = new PurchaseFacade();
purchaseFacade.purchaseItem("iPhone", 1000);

/*** 

En este ejemplo, el patrón Facade nos permite proporcionar una interfaz simplificada para realizar una compra. 
La fachada envuelve los subsistemas de inventario, carrito de compras y procesamiento de pagos, 
ocultando su complejidad y proporcionando una única interfaz fácil de usar para los clientes.

*/

File: 02. Estructurales/05. Facade/01. facade.py
# Definimos nuestro subsistema de inventario
class InventorySystem:
    def check_availability(self, item):
        print(f"Verificando disponibilidad de {item} en inventario...")
        # Lógica para verificar la disponibilidad en el inventario
        return True

# Definimos nuestro subsistema de carrito de compras
class ShoppingCart:
    def add_item(self, item):
        print(f"Agregando {item} al carrito de compras...")
        # Lógica para agregar un elemento al carrito de compras

    def remove_item(self, item):
        print(f"Eliminando {item} del carrito de compras...")
        # Lógica para eliminar un elemento del carrito de compras

# Definimos nuestro subsistema de procesamiento de pagos
class PaymentProcessor:
    def process_payment(self, amount):
        print(f"Procesando pago de {amount}...")
        # Lógica para procesar el pago
        return True

# Definimos nuestra fachada que envuelve los subsistemas
class PurchaseFacade:
    def __init__(self):
        self.inventory_system = InventorySystem()
        self.shopping_cart = ShoppingCart()
        self.payment_processor = PaymentProcessor()

    def purchase_item(self, item, amount):
        if self.inventory_system.check_availability(item):
            self.shopping_cart.add_item(item)
            if self.payment_processor.process_payment(amount):
                print(f"Compra exitosa: {item}")
            else:
                print("Error al procesar el pago")
        else:
            print("El artículo no está disponible en inventario")

# Utilizamos el patrón Facade para realizar una compra de manera simplificada
purchase_facade = PurchaseFacade()
purchase_facade.purchase_item("iPhone", 1000)